gap/fp/tietze.gi

```diff
@@ -0,0 +1,1636 @@
+############################################################################
+##
+##  tietze.gi
+##  Copyright (C) 2021                                    Tom Conti-Leslie
+##                                                              Ben Spiers
+##
+##  Licensing information can be found in the README file of this package.
+##
+############################################################################
+##
+
+########################################################################
+# This file is organised as follows:
+#
+# 1.  Definition of the Semigroup Tietze (IsStzPresentation) object
+# 2.  Viewing methods for IsStzPresentation objects
+# 3.  Internal Tietze transformation functions
+# 4.  User-accessible Tietze transformation functions (wrappers)
+# 5.  Internal helper functions for word/relation manipulation etc.
+# 6.  Internal auto-checkers and appliers for presentation simplifying
+# 7.  SimplifyPresentation etc.
+# 8.  Other
+########################################################################
+
+########################################################################
+# 1. Definition of the Semigroup Tietze (IsStzPresentation) object
+########################################################################
+
+InstallMethod(StzPresentation, "for a finitely presented semigroup",
+[IsFpSemigroup],
+function(S)
+  local gens, out, rels, type;
+
+  type := NewType(NewFamily("StzFamily", IsStzPresentation),
+                  IsStzPresentation and IsComponentObjectRep);
+
+  rels := List(RelationsOfFpSemigroup(S),
+              x -> [LetterRepAssocWord(x[1]), LetterRepAssocWord(x[2])]);
+  gens := List(GeneratorsOfSemigroup(S), x -> ViewString(x));
+
+  out := rec(gens               := gens,
+             rels               := rels,
+             unreducedSemigroup := S,
+             tietzeForwardMap   := List([1 .. Length(GeneratorsOfSemigroup(S))],
+                                        x -> [x]),
+             tietzeBackwardMap  := List([1 .. Length(GeneratorsOfSemigroup(S))],
+                                        x -> [x]),
+             usedGens           := Set(gens));
+
+  return ObjectifyWithAttributes(out, type,
+                                  RelationsOfStzPresentation,
+                                  out!.rels,
```

Comment:
I think to avoid duplication, you might want to rename these `out!.rels` -> `out!.RelationsOfStzPresentation`. Also if `type` doesn't belong to `IsAttributeStoringRep`, then does `ObjectifyWithAttributes` do anything?